rtui2/widgets/graph_panel.py

```diff
@@ -0,0 +1,159 @@
+from __future__ import annotations
+
+from rich.text import Text
+from textual.app import ComposeResult
+from textual.events import Key
+from textual.widgets import Static, Tree
+from textual.widgets.tree import TreeNode
+
+from ..ros import RosClient, RosEntity, RosEntityType
+from ..ros.dependency_graph import RosDependencyGraph, RosDependencyNode
+
+EXPANSION_DEPTH = 2
+
+
+class TreeLabel:
+    @staticmethod
+    def label(entity: RosEntity) -> Text:
+        style = ""
+        if entity.type == RosEntityType.Node:
+            style = "green"
+        elif entity.type == RosEntityType.Topic:
+            style = "white"
+        return Text(entity.name, style=style)
+
+    @staticmethod
+    def leaf_label(entity: RosEntity) -> Text:
+        text = TreeLabel.label(entity)
+        text.stylize("bold underline")
+        return text
+
+    @staticmethod
+    def error(msg: str) -> Text:
+        return Text(f"ERROR: {msg}", style="red")
+
+    NO_PUBLISHER = Text("[No publisher]", style="yellow")
+
+    @staticmethod
+    def is_no_publisher(label: str | Text) -> bool:
+        plain = label.plain if isinstance(label, Text) else str(label)
+        return plain.strip() == "No publisher"
+
+
+class RosEntityGraphPanel(Static):
+    """ROSエンティティの依存関係をツリー表示するパネル"""
+
+    def __init__(
+        self, ros: RosClient, entity: RosEntity | None = None, **kwargs
+    ) -> None:
+        super().__init__(**kwargs)
+        self._ros = ros
+        self._entity = entity
+        self._tree: Tree[RosEntity] | None = None
+
+    def compose(self) -> ComposeResult:
+        if self._tree:
+            yield self._tree
+
+    def set_entity(self, entity: RosEntity) -> None:
+        if entity != self._entity:
+            self._entity = entity
+            self.update_graph()
+
+    def update_graph(self) -> None:
+        if self._tree:
+            self._tree.remove()
+
+        if self._entity is None:
+            return
+
+        self._tree = Tree(TreeLabel.label(self._entity), data=self._entity)
+        self.mount(self._tree)
+
+        graph = RosDependencyGraph(self._entity, self._ros, max_depth=EXPANSION_DEPTH)
+        self._populate_tree(self._tree.root, graph.root)
+        self._tree.root.expand()
```

Comment:
The node expansion logic is duplicated between `on_tree_node_selected()` and `update_graph()`. Both methods follow the same pattern:

  1. Create `RosDependencyGraph`
  2. Call `_populate_tree()`
  3. Expand the node

How about creating a helper method like `_rebuild_node()` to reduce duplication?

```suggestion
    def update_graph(self) -> None:
        if self._tree:
            self._tree.remove()

        if self._entity is None:
            return

        self._tree = Tree(TreeLabel.label(self._entity), data=self._entity)
        self.mount(self._tree)
        self._rebuild_node(self._tree.root, self._entity)
```

Reply:
I have consolidate the node expansion logic in `_populate_tree()` https://github.com/whill-labs/rtui2/pull/2/commits/340d4ab7de0fbe72d8664b85dbdbb2fd61094dad